File: CS/CustomRangeTrackBar.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Repository;
using System.ComponentModel;
using DevExpress.XtraEditors.Drawing;
using DevExpress.Utils;
using DevExpress.XtraEditors.ViewInfo;
using System.Drawing;
using DevExpress.XtraEditors.Controls;

namespace CustomRangeTrackBarControl
{
    [System.ComponentModel.DesignerCategory("None")]
    class MultipleRangeTrackBar : RangeTrackBarControl
    {
        #region Initialization
        static MultipleRangeTrackBar()
        {
            RepositoryItemMultipleRangeTrackBar.Register();
        }
        
        public override string EditorTypeName
        {
            get { return RepositoryItemMultipleRangeTrackBar.EditorName; }
        }
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemMultipleRangeTrackBar Properties
        {
            get { return base.Properties as RepositoryItemMultipleRangeTrackBar; }
        }

        internal new MultiRangeViewInfo ViewInfo
        {
            get { return base.ViewInfo as MultiRangeViewInfo; }
        }
        #endregion

        public MultipleRangeTrackBar()
        {
            if (innerRanges == null)  innerRanges = new List<TrackBarRange>();
        }

        void RefreshInnerRanges()
        {
            RangeList list = (RangeList)EditValue;
            innerRanges.Clear();
            for (int i = 0; i < list.Count; i++)
                innerRanges.Add(list.GetValue(i));
        }

        bool lockEditValueChanged = false;
        protected override void OnEditValueChanged()
        {
            if (!lockEditValueChanged)
            {
                base.OnEditValueChanged();
                if (innerRanges == null) innerRanges = new List<TrackBarRange>();
                RangeList list = EditValue as RangeList;
                RefreshInnerRanges();
            }
        }

        bool IsEqualEditValue(RangeList list)
        {
            RangeList editVal = EditValue as RangeList;
            if (editVal.Count != list.Count) return false;
            for (int i = 0; i < list.Count; i++)
                if (editVal.GetValue(i).Minimum != list.GetValue(i).Minimum || editVal.GetValue(i).Maximum != list.GetValue(i).Maximum)
                    return false;
            return true;
        }

        protected override void OnEditValueChanging(DevExpress.XtraEditors.Controls.ChangingEventArgs e)
        {
            lockEditValueChanged = true;
            base.OnEditValueChanging(e);
            lockEditValueChanged = false;
            if (e.Cancel == false && !IsEqualEditValue((e.NewValue as RangeList)))
            {
                editVal = e.NewValue as RangeList;
                RefreshInnerRanges();
                OnEditValueChanged();
            }
        }

        void OnEditValueChangingCore(RangeList oldVal, RangeList newVal)
        {
            ChangingEventArgs e = new ChangingEventArgs(oldVal, newVal);
            OnEditValueChanging(e);
        }

        void CreateEditValue()
        {
            editVal = new RangeList();
            editVal.ListChanged += new EventHandler(editVal_ListChanged);
        }

        void editVal_ListChanged(object sender, EventArgs e)
        {
            Refresh();
        }

        RangeList editVal;
        public override object EditValue
        {
            get
            {
                return editVal;
            }
            set
            {
                if (editVal == null) CreateEditValue();
                if (value.GetType() == typeof(TrackBarRange))
                {
                    editVal.ChangeValue(0, (TrackBarRange)value);
                    OnEditValueChanged();
                } else
                    if (value.GetType() == typeof(RangeList))
                    {
                        OnEditValueChangingCore(editVal, value as RangeList);
                        editVal.ListChanged += new EventHandler(editVal_ListChanged);
                        Refresh();
                    }
            }
        }
        #region Add, Remove, Get, Change

        bool CheckAdd(int maximum)
        {
            foreach (TrackBarRange range in innerRanges)
                if (range.Minimum <= maximum) return false;
            return true;
        }

        TrackBarRange ConvertItem(TrackBarRange item)
        {
            TrackBarRange range = item;
            range.Minimum = Math.Max(item.Minimum, Properties.Minimum);
            range.Maximum = Math.Min(item.Maximum, Properties.Maximum);
            return range;
        }

        void AddItem(TrackBarRange item)
        {
            TrackBarRange range = ConvertItem(item);
            RangeList oldVal = new RangeList((RangeList)EditValue);
            RangeList newVal = new RangeList(oldVal); 
            newVal.Add(item);
            OnEditValueChangingCore(oldVal, newVal);
        }

        void ChangeItem(int index, TrackBarRange item)
        {
            TrackBarRange range = ConvertItem(item);
            if (innerRanges[index].Maximum != item.Maximum || innerRanges[index].Minimum != item.Minimum)
            {
                RangeList oldVal = new RangeList((RangeList)EditValue);
                RangeList newVal = new RangeList(oldVal); 
                newVal.ChangeValue(index ,item);
                OnEditValueChangingCore(oldVal, newVal);
            }
        }

        void RemoveItem(int index)
        {
            RangeList oldVal = new RangeList((RangeList)EditValue);
            RangeList newVal = new RangeList(oldVal); 
            newVal.RemoveAt(index);
            OnEditValueChangingCore(oldVal, newVal);
        }

        public void AddNewRange()
        {
            AddNewRange(0, 0);
        }

        public void AddNewRange(int minimum, int maximum)
        {
            TrackBarRange item = new TrackBarRange(minimum, maximum);
            if (CheckAdd(item.Maximum))
                AddItem(item);
        }

        public void RemoveRange(int index)
        {
            if (index > 0 && innerRanges.Count > index)
                RemoveItem(index);
        }

        public TrackBarRange GetValue(int index)
        {
            return innerRanges[index];
        }

        public void ChangeValue(TrackBarRange range, int index)
        {
            if (innerRanges.Count > index)
                ChangeItem(index, range);
        }

        public void ChangeValue(int minimum, int maximum, int index)
        {
            TrackBarRange range = new TrackBarRange(minimum, maximum);
            ChangeValue(range, index);
        }
        #endregion

        #region Intersection algorithm
        public delegate void IntersectEventHandler(IntersectEventArgs e );
        public event IntersectEventHandler Intersect;

        void OnIntersect(ThumbType thumbType_, int draggedThumb_, int value_)
        {
            if (Intersect != null)
            {
                IntersectEventArgs e = new IntersectEventArgs(thumbType_, draggedThumb_, value_);
                Intersect(e);
            }
        }

        bool CheckMinChange(int startIndex ,int newMinValue, bool raiseEvent)
        {
            bool check = true;
            int thumb = 0;
            if (newMinValue < Properties.Minimum) { check = false; raiseEvent = false; }
            for (int i = startIndex; i < innerRanges.Count; i++)
            {
                if (innerRanges[i].Maximum >= newMinValue)
                {
                    check = false;
                    thumb = i;
                    break;
                }
            }
            if (!check && raiseEvent)
                OnIntersect(ThumbType.Minimum, thumb - 1, innerRanges[thumb].Maximum);
            return check;
        }

        bool CheckMaxChange(int lastIndex, int newMaxValue, bool raiseEvent)
        {
            bool check = true;
            int thumb = 0;
            if (newMaxValue > Properties.Maximum) { check = false; raiseEvent = false; }
            for (int i = 0; i < lastIndex; i++)
            {
                if (innerRanges[i].Minimum <= newMaxValue)
                {
                    check = false;
                    thumb = i;
                    break;
                }
            }
            if (!check && raiseEvent)
                OnIntersect(ThumbType.Maximum, thumb + 1, innerRanges[thumb].Minimum);
            return check;
        }
        #endregion

        protected virtual TrackBarRange ChangeMinThumbPosition(int dragThumb, int newMinValue)
        {
            TrackBarRange range = innerRanges[dragThumb];
            if (CheckMinChange(ViewInfo.DraggedThumb + 1, newMinValue, true))
                range = new TrackBarRange(newMinValue, innerRanges[ViewInfo.DraggedThumb].Maximum);
            return range;
        }

        protected virtual TrackBarRange ChangeMaxThumbPosition(int dragThumb, int newMaxValue)
        {
            TrackBarRange range = innerRanges[dragThumb];
            if (CheckMaxChange(ViewInfo.DraggedThumb, newMaxValue, true))
                range = new TrackBarRange(innerRanges[ViewInfo.DraggedThumb].Minimum, newMaxValue);
            return range;
        }

        protected override void OnMouseMove(System.Windows.Forms.MouseEventArgs e)
        {
            DXMouseEventArgs ee = DXMouseEventArgs.GetMouseArgs(e);
            if (ee.Handled) return;
            if (Properties.ReadOnly) return;
            if (ViewInfo.PressedInfo.HitTest == EditHitTest.Button)
            {
                ChangeItem(ViewInfo.DraggedThumb, ChangeMinThumbPosition(ViewInfo.DraggedThumb, Math.Min(ViewInfo.ValueFromPoint(ViewInfo.ControlToClient(new Point(e.X, e.Y))),
                    innerRanges[ViewInfo.DraggedThumb].Maximum)));
                ShowMinValue();
            }
            else if (ViewInfo.PressedInfo.HitTest == EditHitTest.Button2)
            {
                ChangeItem(ViewInfo.DraggedThumb, ChangeMaxThumbPosition(ViewInfo.DraggedThumb, Math.Max(innerRanges[ViewInfo.DraggedThumb].Minimum,
                    ViewInfo.ValueFromPoint(ViewInfo.ControlToClient(new Point(e.X, e.Y))))));
                ShowMaxValue();
            }
        }

        protected override void ShowMaxValue()
        {
            ShowValue(innerRanges[ViewInfo.DraggedThumb].ToString(), ViewInfo.MaxThumbBoundses[ViewInfo.DraggedThumb]);
        }

        protected override void ShowMinValue()
        {
            ShowValue(innerRanges[ViewInfo.DraggedThumb].ToString(), ViewInfo.MinThumbBoundses[ViewInfo.DraggedThumb]);
        }

        int GetClosestThumb(int val)
        {
            int value = 0;
            int diff = Int32.MaxValue;
            for (int i = 0; i < innerRanges.Count; i++)
            {
                if (Math.Abs(val - innerRanges[i].Minimum) < diff)
                {
                    value = i * 2;
                    diff = Math.Abs(val - innerRanges[i].Minimum);
                }
                if (Math.Abs(val - innerRanges[i].Maximum) <= diff)
                {
                    value = i * 2 + 1;
                    diff = Math.Abs(val - innerRanges[i].Maximum);
                }
            }
            return value;
        }

        protected override void UpdateValueFromPoint(Point pt)
        {
            if (ViewInfo.DraggedThumb < 0)
            {
                if (!ShouldUpdateValueFromPoint(pt)) return;
                int value = ViewInfo.ValueFromPoint(ViewInfo.ControlToClient(pt));
                int val = GetClosestThumb(value);
                if (val % 2 == 0) ChangeValue(new TrackBarRange(value, innerRanges[val / 2].Maximum), val / 2);
                else ChangeValue(new TrackBarRange(innerRanges[val / 2].Minimum, value), val / 2);
            }
        }

        protected override void OnMouseUp(System.Windows.Forms.MouseEventArgs e)
        {
            base.OnMouseUp(e);
            ViewInfo.DraggedThumb = -1;
        }
                
        private List<TrackBarRange> innerRanges;

        public int ThumbsCount { get { return innerRanges.Count; } }
    }
}

File: VB/CustomRangeTrackBar.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Repository
Imports System.ComponentModel
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.Utils
Imports DevExpress.XtraEditors.ViewInfo
Imports System.Drawing
Imports DevExpress.XtraEditors.Controls

Namespace CustomRangeTrackBarControl
	<System.ComponentModel.DesignerCategory("None")> _
	Friend Class MultipleRangeTrackBar
		Inherits RangeTrackBarControl
		#Region "Initialization"
		Shared Sub New()
			RepositoryItemMultipleRangeTrackBar.Register()
		End Sub

		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return RepositoryItemMultipleRangeTrackBar.EditorName
			End Get
		End Property
		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
		Public Shadows ReadOnly Property Properties() As RepositoryItemMultipleRangeTrackBar
			Get
				Return TryCast(MyBase.Properties, RepositoryItemMultipleRangeTrackBar)
			End Get
		End Property

		Friend Shadows ReadOnly Property ViewInfo() As MultiRangeViewInfo
			Get
				Return TryCast(MyBase.ViewInfo, MultiRangeViewInfo)
			End Get
		End Property
		#End Region

		Public Sub New()
			If innerRanges Is Nothing Then
				innerRanges = New List(Of TrackBarRange)()
			End If
		End Sub

		Private Sub RefreshInnerRanges()
			Dim list As RangeList = CType(EditValue, RangeList)
			innerRanges.Clear()
			For i As Integer = 0 To list.Count - 1
				innerRanges.Add(list.GetValue(i))
			Next i
		End Sub

		Private lockEditValueChanged As Boolean = False
		Protected Overrides Sub OnEditValueChanged()
			If (Not lockEditValueChanged) Then
				MyBase.OnEditValueChanged()
				If innerRanges Is Nothing Then
					innerRanges = New List(Of TrackBarRange)()
				End If
				Dim list As RangeList = TryCast(EditValue, RangeList)
				RefreshInnerRanges()
			End If
		End Sub

		Private Function IsEqualEditValue(ByVal list As RangeList) As Boolean
			Dim editVal As RangeList = TryCast(EditValue, RangeList)
			If editVal.Count <> list.Count Then
				Return False
			End If
			For i As Integer = 0 To list.Count - 1
				If editVal.GetValue(i).Minimum <> list.GetValue(i).Minimum OrElse editVal.GetValue(i).Maximum <> list.GetValue(i).Maximum Then
					Return False
				End If
			Next i
			Return True
		End Function

		Protected Overrides Sub OnEditValueChanging(ByVal e As DevExpress.XtraEditors.Controls.ChangingEventArgs)
			lockEditValueChanged = True
			MyBase.OnEditValueChanging(e)
			lockEditValueChanged = False
			If e.Cancel = False AndAlso (Not IsEqualEditValue((TryCast(e.NewValue, RangeList)))) Then
				editVal = TryCast(e.NewValue, RangeList)
				RefreshInnerRanges()
				OnEditValueChanged()
			End If
		End Sub

		Private Sub OnEditValueChangingCore(ByVal oldVal As RangeList, ByVal newVal As RangeList)
			Dim e As New ChangingEventArgs(oldVal, newVal)
			OnEditValueChanging(e)
		End Sub

		Private Sub CreateEditValue()
			editVal = New RangeList()
			AddHandler editVal.ListChanged, AddressOf editVal_ListChanged
		End Sub

		Private Sub editVal_ListChanged(ByVal sender As Object, ByVal e As EventArgs)
			Refresh()
		End Sub

		Private editVal As RangeList
		Public Overrides Property EditValue() As Object
			Get
				Return editVal
			End Get
			Set(ByVal value As Object)
				If editVal Is Nothing Then
					CreateEditValue()
				End If
				If value.GetType() Is GetType(TrackBarRange) Then
					editVal.ChangeValue(0, CType(value, TrackBarRange))
					OnEditValueChanged()
				Else
					If value.GetType() Is GetType(RangeList) Then
						OnEditValueChangingCore(editVal, TryCast(value, RangeList))
						AddHandler editVal.ListChanged, AddressOf editVal_ListChanged
						Refresh()
					End If
				End If
			End Set
		End Property
		#Region "Add, Remove, Get, Change"

		Private Function CheckAdd(ByVal maximum As Integer) As Boolean
			For Each range As TrackBarRange In innerRanges
				If range.Minimum <= maximum Then
					Return False
				End If
			Next range
			Return True
		End Function

		Private Function ConvertItem(ByVal item As TrackBarRange) As TrackBarRange
			Dim range As TrackBarRange = item
			range.Minimum = Math.Max(item.Minimum, Properties.Minimum)
			range.Maximum = Math.Min(item.Maximum, Properties.Maximum)
			Return range
		End Function

		Private Sub AddItem(ByVal item As TrackBarRange)
			Dim range As TrackBarRange = ConvertItem(item)
			Dim oldVal As New RangeList(CType(EditValue, RangeList))
			Dim newVal As New RangeList(oldVal)
			newVal.Add(item)
			OnEditValueChangingCore(oldVal, newVal)
		End Sub

		Private Sub ChangeItem(ByVal index As Integer, ByVal item As TrackBarRange)
			Dim range As TrackBarRange = ConvertItem(item)
			If innerRanges(index).Maximum <> item.Maximum OrElse innerRanges(index).Minimum <> item.Minimum Then
				Dim oldVal As New RangeList(CType(EditValue, RangeList))
				Dim newVal As New RangeList(oldVal)
				newVal.ChangeValue(index,item)
				OnEditValueChangingCore(oldVal, newVal)
			End If
		End Sub

		Private Sub RemoveItem(ByVal index As Integer)
			Dim oldVal As New RangeList(CType(EditValue, RangeList))
			Dim newVal As New RangeList(oldVal)
			newVal.RemoveAt(index)
			OnEditValueChangingCore(oldVal, newVal)
		End Sub

		Public Sub AddNewRange()
			AddNewRange(0, 0)
		End Sub

		Public Sub AddNewRange(ByVal minimum As Integer, ByVal maximum As Integer)
			Dim item As New TrackBarRange(minimum, maximum)
			If CheckAdd(item.Maximum) Then
				AddItem(item)
			End If
		End Sub

		Public Sub RemoveRange(ByVal index As Integer)
			If index > 0 AndAlso innerRanges.Count > index Then
				RemoveItem(index)
			End If
		End Sub

		Public Function GetValue(ByVal index As Integer) As TrackBarRange
			Return innerRanges(index)
		End Function

		Public Sub ChangeValue(ByVal range As TrackBarRange, ByVal index As Integer)
			If innerRanges.Count > index Then
				ChangeItem(index, range)
			End If
		End Sub

		Public Sub ChangeValue(ByVal minimum As Integer, ByVal maximum As Integer, ByVal index As Integer)
			Dim range As New TrackBarRange(minimum, maximum)
			ChangeValue(range, index)
		End Sub
		#End Region

		#Region "Intersection algorithm"
		Public Delegate Sub IntersectEventHandler(ByVal e As IntersectEventArgs)
		Public Event Intersect As IntersectEventHandler

		Private Sub OnIntersect(ByVal thumbType_ As ThumbType, ByVal draggedThumb_ As Integer, ByVal value_ As Integer)
			If IntersectEvent IsNot Nothing Then
				Dim e As New IntersectEventArgs(thumbType_, draggedThumb_, value_)
				RaiseEvent Intersect(e)
			End If
		End Sub

		Private Function CheckMinChange(ByVal startIndex As Integer, ByVal newMinValue As Integer, ByVal [raiseEvent] As Boolean) As Boolean
			Dim check As Boolean = True
			Dim thumb As Integer = 0
			If newMinValue < Properties.Minimum Then
				check = False
				[raiseEvent] = False
			End If
			For i As Integer = startIndex To innerRanges.Count - 1
				If innerRanges(i).Maximum >= newMinValue Then
					check = False
					thumb = i
					Exit For
				End If
			Next i
			If (Not check) AndAlso [raiseEvent] Then
				OnIntersect(ThumbType.Minimum, thumb - 1, innerRanges(thumb).Maximum)
			End If
			Return check
		End Function

		Private Function CheckMaxChange(ByVal lastIndex As Integer, ByVal newMaxValue As Integer, ByVal [raiseEvent] As Boolean) As Boolean
			Dim check As Boolean = True
			Dim thumb As Integer = 0
			If newMaxValue > Properties.Maximum Then
				check = False
				[raiseEvent] = False
			End If
			For i As Integer = 0 To lastIndex - 1
				If innerRanges(i).Minimum <= newMaxValue Then
					check = False
					thumb = i
					Exit For
				End If
			Next i
			If (Not check) AndAlso [raiseEvent] Then
				OnIntersect(ThumbType.Maximum, thumb + 1, innerRanges(thumb).Minimum)
			End If
			Return check
		End Function
		#End Region

		Protected Overridable Function ChangeMinThumbPosition(ByVal dragThumb As Integer, ByVal newMinValue As Integer) As TrackBarRange
			Dim range As TrackBarRange = innerRanges(dragThumb)
			If CheckMinChange(ViewInfo.DraggedThumb + 1, newMinValue, True) Then
				range = New TrackBarRange(newMinValue, innerRanges(ViewInfo.DraggedThumb).Maximum)
			End If
			Return range
		End Function

		Protected Overridable Function ChangeMaxThumbPosition(ByVal dragThumb As Integer, ByVal newMaxValue As Integer) As TrackBarRange
			Dim range As TrackBarRange = innerRanges(dragThumb)
			If CheckMaxChange(ViewInfo.DraggedThumb, newMaxValue, True) Then
				range = New TrackBarRange(innerRanges(ViewInfo.DraggedThumb).Minimum, newMaxValue)
			End If
			Return range
		End Function

		Protected Overrides Sub OnMouseMove(ByVal e As System.Windows.Forms.MouseEventArgs)
			Dim ee As DXMouseEventArgs = DXMouseEventArgs.GetMouseArgs(e)
			If ee.Handled Then
				Return
			End If
			If Properties.ReadOnly Then
				Return
			End If
			If ViewInfo.PressedInfo.HitTest = EditHitTest.Button Then
				ChangeItem(ViewInfo.DraggedThumb, ChangeMinThumbPosition(ViewInfo.DraggedThumb, Math.Min(ViewInfo.ValueFromPoint(ViewInfo.ControlToClient(New Point(e.X, e.Y))), innerRanges(ViewInfo.DraggedThumb).Maximum)))
				ShowMinValue()
			ElseIf ViewInfo.PressedInfo.HitTest = EditHitTest.Button2 Then
				ChangeItem(ViewInfo.DraggedThumb, ChangeMaxThumbPosition(ViewInfo.DraggedThumb, Math.Max(innerRanges(ViewInfo.DraggedThumb).Minimum, ViewInfo.ValueFromPoint(ViewInfo.ControlToClient(New Point(e.X, e.Y))))))
				ShowMaxValue()
			End If
		End Sub

		Protected Overrides Sub ShowMaxValue()
			ShowValue(innerRanges(ViewInfo.DraggedThumb).ToString(), ViewInfo.MaxThumbBoundses(ViewInfo.DraggedThumb))
		End Sub

		Protected Overrides Sub ShowMinValue()
			ShowValue(innerRanges(ViewInfo.DraggedThumb).ToString(), ViewInfo.MinThumbBoundses(ViewInfo.DraggedThumb))
		End Sub

		Private Function GetClosestThumb(ByVal val As Integer) As Integer
			Dim value As Integer = 0
			Dim diff As Integer = Int32.MaxValue
			For i As Integer = 0 To innerRanges.Count - 1
				If Math.Abs(val - innerRanges(i).Minimum) < diff Then
					value = i * 2
					diff = Math.Abs(val - innerRanges(i).Minimum)
				End If
				If Math.Abs(val - innerRanges(i).Maximum) <= diff Then
					value = i * 2 + 1
					diff = Math.Abs(val - innerRanges(i).Maximum)
				End If
			Next i
			Return value
		End Function

		Protected Overrides Sub UpdateValueFromPoint(ByVal pt As Point)
			If ViewInfo.DraggedThumb < 0 Then
				If (Not ShouldUpdateValueFromPoint(pt)) Then
					Return
				End If
				Dim value As Integer = ViewInfo.ValueFromPoint(ViewInfo.ControlToClient(pt))
				Dim val As Integer = GetClosestThumb(value)
				If val Mod 2 = 0 Then
					ChangeValue(New TrackBarRange(value, innerRanges(val \ 2).Maximum), val \ 2)
				Else
					ChangeValue(New TrackBarRange(innerRanges(val \ 2).Minimum, value), val \ 2)
				End If
			End If
		End Sub

		Protected Overrides Sub OnMouseUp(ByVal e As System.Windows.Forms.MouseEventArgs)
			MyBase.OnMouseUp(e)
			ViewInfo.DraggedThumb = -1
		End Sub

		Private innerRanges As List(Of TrackBarRange)

		Public ReadOnly Property ThumbsCount() As Integer
			Get
				Return innerRanges.Count
			End Get
		End Property
	End Class
End Namespace
